text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,161 @@
+
+### Protobuf design
+
+```proto
+message RawCoprocessorRequest {
+    kvrpcpb.Context context = 1;
+
+    string copr_name = 2;
+    string copr_version_constraint = 3;
+
+    bytes data = 4;
+}
+
+message RawCoprocessorResponse {
+    bytes data = 1;
+
+    errorpb.Error region_error = 2;
+    string other_error = 3;
+}
+```
+
+### API design
+
+To reduce the learning overhead, it'll be better that the API of the coprocessor plugin get closer to the client. Thus, it'll looks like the `RawClient` in the [Rust Client](https://github.com/tikv/client-rust) with extra txn-like methods e.g. `commit` and `lock`.
+
+```rust
+use std::ops::Range;
+
+pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+pub type KvPair = (Key, Value);
+
+#[derive(Debug)]
+pub struct Region {
+    id: u64,
+    start_key: Key,
+    end_key: Key,
+    region_epoch: RegionEpoch,
+}
+
+#[derive(Debug)]
+pub struct RegionEpoch {
+    pub conf_ver: u64,
+    pub version: u64,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    KeyNotInRegion { key: Key, region: Region },
+    // More
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[async_trait]
+pub trait RawStorage: Send {
+    async fn get(&self, key: Key) -> Result<Option<Value>>;
+    async fn batch_get(&self, keys: Vec<Key>) -> Result<Vec<KvPair>>;
+    async fn scan(&self, key_range: Range<Key>) -> Result<Vec<Value>>;
+    async fn put(&mut self, key: Key, value: Value) -> Result<()>;
+    async fn batch_put(&mut self, kv_pairs: Vec<KvPair>) -> Result<()>;
+    async fn delete(&mut self, key: Key) -> Result<()>;
+    async fn batch_delete(&mut self, keys: Vec<Key>) -> Result<()>;
+    async fn delete_range(&mut self, key_range: Range<Key>) -> Result<()>;
+}
+
+pub trait Coprocessor: Send + Sync {
+    fn on_raw_coprocessor_request(
+        &self,
+        region: Region,
+        request: Vec<u8>,
+        storage: Box<dyn RawStorage>,
+    ) -> Result<Vec<u8>>;
+```
+
+### Keyspace
+
+Keyspace[[RFC]](https://github.com/tikv/rfcs/pull/39)[[The most updated design doc]](https://docs.google.com/document/d/1x17-urAqToDo8TVXJroEHtc76fdssFaoANjSaNDhjKg/edit) is an incoming feature of TiKV that is highly related to coprocessor plugin. Keyspace determines whether a range of key should only be used in transaction mode or in RawKV mode. Since coprocessor works in either RawKV mode or txn mode, surely coprocessor plugin framework should aware of Keyspace. The details is TBD.
+
+## Future work
+
+### Official coprocessor
+
+Provide an official coprocessor plugin which defines common primitives like `bool`, `f32`, `UTF8String`, `Array<T>` along with codec for them, moreover, implements simple scanning expressions or even aggregations.
```

Comment:
Why not just define them in SDK?

Reply:
First of all, encoding of `bool`, `f32`, `UTF8String`, `Array<T>` depends on the library which plugin uses (e.g. protobuf, Apache Arrow, etc.). At the same time, different plugins may process and interpret raw byte data from clients in different ways. From my perspective, it is the plugins' job to define how to interpret raw byte data, and what types they could process. With this design, SDK only needs to provide byte data, and the types should be specified in plugin itself.

Maybe someday we would find out that all plugins are using these same set of types, and lift those types from the plugin to the SDK?